### ISSUE TITLE
Bug - Core 4.0.8 and Service Dependency

### ISSUE THREAD

**Issue by box293** [NagiosEnterprises/nagioscore#19]
This was reported in the forums and affects Core 4.0.8.

http://support.nagios.com/forum/viewtopic.php?f=16&t=29205&p=112139#p112139

Problem:

When hostgroups are used for the hostgroup_name directive, the service dependency appears to be ignored. This does not matter if the host's are assigned to the hostgroup via their host object definition OR inside the hostgroup definition.

 

When I say ignored, when the Master service goes down, core continues to check the services depending on it. If these are down then notifications are sent.

When I tail nagios.log I can see the notifications.

When using Core 3.5.1, when the Master service goes down, core DOES NOT continue to check the services depending on it. The next check time keeps getting pushed back.

The following config settings can be used to reproduce the issue:

define host{
        use                     linux-server
        host_name               host1
        alias                   host1
        address                 127.0.0.1
        hostgroups              my_test
        }

define host{
        use                     linux-server
        host_name               host2
        alias                   host2
        address                 127.0.0.1
        hostgroups              my_test
        }

define command{
        command_name    check-host-alive-specific
        command_line    $USER1$/check_ping -H $ARG1$ -w 3000.0,80% -c 5000.0,100% -p 5
        }

define service{
        use                             local-service
        host_name                       host1
        service_description             master
        check_command                   check-host-alive-specific!192.168.207.142
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define service{
        use                             local-service
        host_name                       host1
        service_description             one
        check_command                   check-host-alive-specific!192.168.207.143
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define service{
        use                             local-service
        host_name                       host1
        service_description             two
        check_command                   check-host-alive-specific!192.168.207.144
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define service{
        use                             local-service
        host_name                       host2
        service_description             master
        check_command                   check-host-alive-specific!192.168.207.130
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define service{
        use                             local-service
        host_name                       host2
        service_description             one
        check_command                   check-host-alive-specific!192.168.207.128
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define service{
        use                             local-service
        host_name                       host2
        service_description             two
        check_command                   check-host-alive-specific!192.168.207.131
        max_check_attempts              2
        check_interval                  1
        retry_interval                  1
        flap_detection_enabled          0
        notification_period             24x7
        notifications_enabled           1
        }

define hostgroup {
        hostgroup_name                          my_test
        alias                                   my_test
        }

When done this way, it works

define servicedependency {
       dependent_host_name                 host1,host2
       dependent_service_description       one,one,two,two
       host_name                          host1,host2
       service_description                 master
       inherits_parent                          1
       execution_failure_criteria               u,c,p,
       notification_failure_criteria            u,c,p,
       dependency_period                        24x7
}

When done this way, it works

define servicedependency {
       dependent_hostgroup_name                 my_test
       dependent_service_description            one,one,two,two
        host_name                          host1,host2
       service_description                      master
       inherits_parent                          1
       execution_failure_criteria               u,c,p,
       notification_failure_criteria            u,c,p,
       dependency_period                        24x7
}

When done this way, it DOES NOT work

define servicedependency {
       dependent_hostgroup_name                 my_test
       dependent_service_description            one,one,two,two
       hostgroup_name                           my_test
       service_description                      master
       inherits_parent                          1
       execution_failure_criteria               u,c,p,
       notification_failure_criteria            u,c,p,
       dependency_period                        24x7
}

**Comment by gwakem** [comment]
I am also experiencing this bug in Nagios Core Version 4.1.0rc1.

**Comment by tmcnag** [comment]
@box293 - I imported your listed configs, as well as the non-working service dependency.  I got these servicedependency listings in objects.cache:

```
define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host1
        dependent_service_description   one
        dependency_period       24x7
        inherits_parent 1
        notification_failure_options    u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host1
        dependent_service_description   one
        dependency_period       24x7
        inherits_parent 1
        execution_failure_options       u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host1
        dependent_service_description   two
        dependency_period       24x7
        inherits_parent 1
        notification_failure_options    u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host1
        dependent_service_description   two
        dependency_period       24x7
        inherits_parent 1
        execution_failure_options       u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host2
        dependent_service_description   one
        dependency_period       24x7
        inherits_parent 1
        notification_failure_options    u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host2
        dependent_service_description   one
        dependency_period       24x7
        inherits_parent 1
        execution_failure_options       u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host2
        dependent_service_description   two
        dependency_period       24x7
        inherits_parent 1
        notification_failure_options    u,c,p
        }

define servicedependency {
        host_name       host2
        service_description     master
        dependent_host_name     host2
        dependent_service_description   two
        dependency_period       24x7
        inherits_parent 1
        execution_failure_options       u,c,p
        }

```

Note a few things.  First, there are duplicates - every even listing is unique and every odd listing is a duplicate (or vice versa, however you want to look at it).  This leads me to believe that this bug is in the config parser/object builder, as opposed to the check-handling logic.

Second, note that none of the `host_name` parameters reference `host1`, only `host2`.  I believe, combined with my first point above, that these are supposed to be `host1` in half of these configs, and that the duplicates are maybe not duplicates in the classic sense of the word, but rather "almost correct".

Conclusion:  Somewhere in the object builder we are either:

* only taking one element out of the hostgroup `my_test` instead of looping over it, or
* properly looping, but overwriting previous correct values with the last element

My bet is on the former, though I am not yet certain how the hostgroup member lists are built up, if it has to do with config file name, host name, or some other factor.

It is also worth mentioning, I tried your two servicedependency configs that you say worked, and got similar duplicates, except that the `host_name` was set to `host1`.  Can you confirm this behavior?

One last point of clarification, is it your intent that all of the `one` and `two` services on all of the hosts in `my_test` are dependent on the `master` service on the respective host?

**Comment by box293** [comment]
> It is also worth mentioning, I tried your two servicedependency configs that you say worked, and got similar duplicates, except that the host_name was set to host1. Can you confirm this behavior?

Yes, I just had a play and I can confirm.



> One last point of clarification, is it your intent that all of the one and two services on all of the hosts in my_test are dependent on the master service on the respective host?

So I went back to the documentation for clarification. 
http://nagios.sourceforge.net/docs/nagioscore/4/en/dependencies.html
> You can create several dependencies for a given service, but you must add a separate service dependency definition for each dependency you create. 

However here:
http://nagios.sourceforge.net/docs/nagioscore/4/en/objectdefinitions.html#servicedependency
> dependent_host_name: 	This directive is used to identify the short name(s) of the host(s) that the dependent service "runs" on or is associated with. 

> dependent_hostgroup_name: 	This directive is used to specify the short name(s) of the hostgroup(s) that the dependent service "runs" on or is associated with. 

> servicegroup_name: 	This directive is used to specify the short name(s) of the servicegroup(s) > that will inherit the dependency. Multiple servicegroups should be separated by commas.
dependent_servicegroup_name: 	This directive is used to specify the short name(s) of the servicegroup(s) that the dependent service "runs" on or is associated with. Multiple servicegroups should be separated by commas. 

> host_name: 	This directive is used to identify the short name(s) of the host(s) that the service that is being depended upon (also referred to as the master service) "runs" on or is associated with. Multiple hosts should be separated by commas.

> hostgroup_name: 	This directive is used to identify the short name(s) of the hostgroup(s) that the service that is being depended upon (also referred to as the master service) "runs" on or is associated with. Multiple hostgroups should be separated by commas. 


So what I am highlighting here is one part of the documentation states "but you must add a separate service dependency definition for each dependency you create" however other parts of the documentation state "Multiple xxx should be separated by commas" which implies that separate service dependencies are not required.

So perhaps Nagios is working as expected and it's just that I'm creating incorrect definitions? 
If this is the case, then perhaps the nagios verification process should detect this and thrown an error.
Also, the documentation should be update if it is not correct or clear.

The other thing which I must make clear is that this problem was originally detected in Nagios XI, and Nagios XI uses CCM to create the flat config files and this is how I originally tested. So perhaps CCM is the original culprit in creating incorrect definitions?


> One last point of clarification, is it your intent that all of the one and two services on all of the hosts in my_test are dependent on the master service on the respective host?

Yes.

What are your thoughts on this?

**Comment by box293** [comment]
So I decided to go back to Core 3.5.1 and re-test. I realised that my original test may have been a little flawed and not as thorough as they should have been.

So here are my revised configs:

```
define host{
use linux-server
host_name host1
alias host1
address 127.0.0.1
hostgroups my_test
}

define host{
use linux-server
host_name host2
alias host2
address 127.0.0.1
hostgroups my_test
}

define host{
use linux-server
host_name host3
alias host3
address 127.0.0.1
hostgroups my_test
}

define command{
command_name check-host-alive-specific
command_line $USER1$/check_ping -H $ARG1$ -w 3000.0,80% -c 5000.0,100% -p 5
}

define service{
use local-service
host_name host1
service_description master
check_command check-host-alive-specific!10.25.254.180
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host1
service_description one
check_command check-host-alive-specific!10.25.5.30
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host1
service_description two
check_command check-host-alive-specific!10.25.5.20
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host2
service_description master
check_command check-host-alive-specific!10.25.14.1
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host2
service_description one
check_command check-host-alive-specific!10.25.5.30
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host2
service_description two
check_command check-host-alive-specific!10.25.5.20
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host3
service_description master
check_command check-host-alive-specific!10.25.254.50
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host3
service_description one
check_command check-host-alive-specific!10.25.5.30
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define service{
use local-service
host_name host3
service_description two
check_command check-host-alive-specific!10.25.5.20
max_check_attempts 2
check_interval 1
retry_interval 1
flap_detection_enabled 0
notification_period 24x7
notifications_enabled 1
}

define hostgroup {
hostgroup_name host1
alias host1
members host1
}

define hostgroup {
hostgroup_name host1_and_host2
alias host1_and_host2
members host1,host2
}

define hostgroup {
hostgroup_name host1_and_host2_and_host3
alias host1_and_host2_and_host3
members host1,host2,host3
}
```

Specifically I've added a third host and the master service on host1, host2 and host3 points to differnt IP addresses. I've also created some additional hostgroups.

Starting off with a basic dependency.

```
define servicedependency {
dependent_host_name host1
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}
```

This all looks good.

Lets add more hosts to the mix.
host1 "master" service has the services "one" and "two" on host1, host2 and host3 being dependant.

In Core 3.5.1 it does not matter if my dependency is defined using the host names or hostgroup names:

```
define servicedependency {
dependent_host_name host1,host2,host3
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_host_name host1,host2,host3
dependent_service_description one,two
hostgroup_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_hostgroup_name host1_and_host2_and_host3
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_hostgroup_name host1_and_host2_and_host3
dependent_service_description one,two
hostgroup_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

In any definitions above, the objects created in objects.cache are identical:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}
```

This all looks good.


Also, a dependency where a master service on host1 and host2 are being depended upon by host3 services one and two.

```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
host_name host1,host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
hostgroup_name host1_and_host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
creates:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	execution_failure_options	u,c,p
	}
```
<br>
So the purpose of this was to identify that it works correctly in core 3.5.1. 
<br>
<br>
Now in Core 4.0.8

Lets start with a simple dependency:
```
define servicedependency {
dependent_host_name host1
dependent_service_description one
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

Which seems to be creating a duplicate object.


Let's add another dependent service:

```
define servicedependency {
dependent_host_name host1
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

Which seems to be creating a duplicate object each time.

So that would be the first bug we can identify.
<br>
<br>
Lets add more hosts to the mix.
host1 "master" service has host1, host2 and host3 with dependent services "one" and "two".

It does not matter if my dependency is defined using the host names or hostgroup names:

```
define servicedependency {
dependent_host_name host1,host2,host3
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_host_name host1,host2,host3
dependent_service_description one,two
hostgroup_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_hostgroup_name host1_and_host2_and_host3
dependent_service_description one,two
host_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```
or
```
define servicedependency {
dependent_hostgroup_name host1_and_host2_and_host3
dependent_service_description one,two
hostgroup_name host1
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

In any definitions above, the objects created in objects.cache are identical:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host1
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host2
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

Which seems to be exhibiting that same behviour of creating a duplicate object each time.

<br>
<br>
Next, a dependency where a master service on host1 and host2 are being depended upon by host3 services one and two.

This example uses host_name
```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
host_name host1,host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:
```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

In this case, none of the host2 objects are created

This example uses hostgroup_name
```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
hostgroup_name host1_and_host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

In this case, none of the host1 objects are created


So just to take it one step further:
A dependency where a master service on host1, host2 and host3 are being depended upon by host3 services one and two.

This example uses host_name

```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
host_name host1,host2,host3
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

In this case, none of the host2 OR host3 objects are created

This example uses hostgroup_name

```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
hostgroup_name host1_and_host2_and_host3
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

creates:

```
define servicedependency {
	host_name	host3
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host3
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host3
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host3
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```

In this case, none of the host1 or host2 objects are created.
I also tried:
hostgroup_name host1_and_host2_and_host3,host1_and_host2
but the same result was produced.


So these appear to be separate bugs.

When multiple hosts are defined by host_name, only the first host in the list objects are created 

When multiple hosts are defined by hostgroup_name, only the last host in the hostgroup have objects are created
<br>
<br>
Summary:
<b>Bug 1:</b>
Duplicate service dependencies are created in objects.cache

<b>Bug 2:</b>
When multiple hosts are defined by host_name, only the first host in the list have objects created in objects.cache

<b>Bug 3:</b>
When multiple hosts are defined by hostgroup_name, only the last host in the hostgroup have objects created in objects.cache

**Comment by jfrickson** [comment]
Bug 1 is actually not a bug. There are two dependencies created for each host/service. One is an EXECUTION_DEPENDENCY and the other is a NOTIFICATION_DEPENDENCY. objects.cache wasn't showing that flag. It is now, however.

Bugs 2 & 3 are actually identical problems, just building the host lists in different orders. The program loops through the hosts and builds service dependencies for each service for each host. (Outer loop = hosts, inner loop = services). The inner loop sets a flag in the service map, so it won't build multiple service dependencies for that host/service. But the outer loops doesn't clear the service map, so no service dependencies get built for the 2nd through nth host.

Test using nagios-maint (https://github.com/NagiosEnterprises/nagioscore/tree/nagios-maint)

**Comment by box293** [comment]
> Bug 1 is actually not a bug. There are two dependencies created for each host/service. One is an EXECUTION_DEPENDENCY and the other is a NOTIFICATION_DEPENDENCY. objects.cache wasn't showing that flag. It is now, however.

Thanks John, you're absolutely right, I must have been like a deer in headlights.

I've tested using 4.1.0rc2 however the bug 2 & 3 behaviour still exists.

```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
host_name host1,host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

Creates:
```
define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host1
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```
In this case, none of the host2 objects are created

This example uses hostgroup_name

```
define servicedependency {
dependent_host_name host3
dependent_service_description one,two
hostgroup_name host1_and_host2
service_description master
inherits_parent 1
execution_failure_criteria u,c,p,
notification_failure_criteria u,c,p,
dependency_period 24x7
}
```

Creates:
```
define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	one
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	notification_failure_options	u,c,p
	}

define servicedependency {
	host_name	host2
	service_description	master
	dependent_host_name	host3
	dependent_service_description	two
	dependency_period	24x7
	inherits_parent	1
	execution_failure_options	u,c,p
	}
```
In this case, none of the host1 objects are created

Also, to confirm the version I am testing with:

Nagios Core 4.1.0rc2
Copyright (c) 2009-present Nagios Core Development Team and Community Contributors
Copyright (c) 1999-2009 Ethan Galstad
Last Modified: 07-17-2015
License: GPL

I cloned the github repository using
```
git clone --recursive https://github.com/NagiosEnterprises/nagioscore.git
```
Will this have cloned the nagios-maint tree? I'm still getting my head around GitHub.

**Comment by jfrickson** [comment]
Clone using
    git clone --progress https://<userid>@github.com/NagiosEnterprises/nagioscore -b nagios-maint

or since you already did a clone, you can do
    git checkout -b nagios-maint origin/nagios-maint

Either should get the correct branch.

**Comment by box293** [comment]
Excellent, now that I've compiled the correct code it works exactly as it should, all the correct dependencies are being created, great work!

**Comment by jfrickson** [comment]
Ok, now that they're all created properly, is the original problem fixed?

**Comment by box293** [comment]
Yes it is. I did some tests today and when the master service goes down, core does not continue to check the services depending on it. The next check time keeps getting pushed back (which is exactly what it should be doing).

Happy days :)

**Comment by jfrickson** [comment]
Fix in nagios-maint. Will be merged into TRUNK at some point after the 4.1.0 release

**Comment by box293** [comment]
I wanted to request that this be included in the 4.1.0 release please.
